test/session_tests/DocumentQueryTest.ts

```diff
@@ -0,0 +1,219 @@
+import {IDocumentStore} from "../../src/Documents/IDocumentStore";
+import {IDocumentSession} from "../../src/Documents/Session/IDocumentSession";
+import {PutIndexesCommand} from "../../src/Database/Commands/PutIndexesCommand";
+import {IndexDefinition} from "../../src/Database/Indexes/IndexDefinition";
+import {IndexFieldOptions} from "../../src/Database/Indexes/IndexFieldOptions";
+import {SortOptions} from "../../src/Database/Indexes/SortOption";
+import {IHash} from "../../src/Utility/Hash";
+
+describe('Document query test', () => {
+  let store: IDocumentStore;
+  let query: IDocumentQuery;
+  let session: IDocumentSession;
+  let defaultDatabase: string, defaultUrl: string;
+
+  beforeEach(function (): void {
+    ({defaultDatabase, defaultUrl} = (this.currentTest as IHash));
+  });
+
+  beforeEach((done: MochaDone) => {
+    store = DocumentStore.create(defaultUrl, defaultDatabase);
+    store.initialize();
+
+    const indexMap: string = [
+      'from doc in docs ',
+      'select new {',
+      'name = doc.name,',
+      'key = doc.key,',
+      'doc_id = doc.key+"_"+doc.name}'
+    ].join('');
+
+
+    const indexSort: IndexDefinition = new IndexDefinition('Testing_Sort', indexMap, null, {
+      fields: {
+        "key": new IndexFieldOptions(SortOptions.Numeric),
+        "doc_id": new IndexFieldOptions(null, null, true)
+      }
+    });
+
+    store.getRequestsExecutor()
+      .execute(new PutIndexesCommand(indexSort))
+      .then((): Promise.Thenable<Object> => {
+        session = store.openSession();
+
+        return Promise.all([
+          session.store(session.create({name: 'test101', key: 2, order: 'a'}, 'product'), 'products/101'),
+          session.store(session.create({name: 'test10', key: 3, order: 'test'}, 'product'), 'products/10'),
+          session.store(session.create({name: 'test106', key: 4, order: 'c'}, 'product'), 'products/106'),
+          session.store(session.create({name: 'test107', key: 5, order: null}, 'product'), 'products/107'),
+          session.store(session.create({name: 'test107', key: 6, order: null}, 'product'), 'products/103'),
+          session.store(session.create({name: 'new_testing', key: 90, order: 'd'}, 'product'), 'products/108'),
+          session.store(session.create({
+            name: 'testing_order',
+            key: 92,
+            product_id: 'products/108'
+          }, 'order'), 'orders/105'),
+          session.store(session.create({
+            name: 'withNesting',
+            product: {name: 'testing_order', key: 4, order: null}
+          }, 'company'), 'company/1')
+        ])
+        .then(() => done());
+      });
+  });
+
+  describe('Index checking', () => {
+    it('should equal dynamic index', (done: MochaDone) => {
+      session = store.openSession();
+      session.query().whereEquals('name', 'test101').get().then((results: Object[]) => {
+        expect(results[0].name).to.equals('test101');
+        done();
```

Comment:
Just a note I believe you can return a Promise in `it()` callback and mocha is going to handle that for you without the need to call `done()`.